cpp/include/IceSSL/Plugin.h

```diff
@@ -401,6 +468,23 @@ class ICESSL_API Certificate :
     virtual bool checkValidity(const IceUtil::Time& t) const = 0;
 #endif
 
+    /**
+     * Returns the value of the key usage extension. The flags <b>KEY_USAGE_DIGITAL_SIGNATURE</b>,
+     * <b>KEY_USAGE_NON_REPUDIATION</b>, <b>KEY_USAGE_KEY_ENCIPHERMENT</b>, <b>KEY_USAGE_DATA_ENCIPHERMENT</b>
+     * <b>KEY_USAGE_KEY_AGREEMENT</b>, <b>KEY_USAGE_KEY_CERT_SIGN</b>, <b>KEY_USAGE_CRL_SIGN</b>,
+     * <b>KEY_USAGE_ENCIPHER_ONLY</b> and <b>KEY_USAGE_DECIPHER_ONLY</b> can be used to check what
+     * key usage bits are set.
+     */
+    virtual unsigned int getKeyUsage() const = 0;
+
+    /**
+     * Returns the value of the extended key usage extension. The flags <b>EXTENDED_KEY_USAGE_ANY_KEY_USAGE</b>,
+     * <b>EXTENDED_KEY_USAGE_SERVER_AUTH</b>, <b>EXTENDED_KEY_USAGE_CLIENT_AUTH</b>,
+     * <b>EXTENDED_KEY_USAGE_CODE_SIGNING</b>, <b>EXTENDED_KEY_USAGE_EMAIL_PROTECTION</b>,
+     * <b>EXTENDED_KEY_USAGE_TIME_STAMPING</b> and <b>EXTENDED_KEY_USAGE_OCSP_SIGNING</b>
+     */
```

Comment:
`can be used to check what extended key use bits are set.`

Reply:
Adding virtual methods to `Certificate` most likely breaks binary compatibility as well.

Reply:
Yes new virtual methods can break binary compatibility, what is the alternative here? 

we can add `int IceSSL::getKeyUsage(const CertificatePtr&)` function and cast the certificate to the implementation class to call the member function.

Reply:
Or a non-virtual `Certificate::getKeyUsage` method that performs the dynamic cast to the concrete implementation? I think this would work too.

Reply:
This doesn't work well with `icesslopenssl` this is the Windows build of the OpenSSL plug-in, the `Certificate` class in `icessl` lib doesn't know about the `OpenSSLCertificateI` implementation in `icesslopenssl` lib.

Reply:
I added a separate class `CertificateExtendedInfo` with these two virtual methods, the different certificate implementations implement this class, I think this is safe as we use a private implementation class for each certificate class. And then the non-virtual methods in `Certificate` cast `this` to this new `CertificateExtendedInfo` class.